File: node/src/vm.ts
import { AttestationResult, makeResult, orderChecks } from "./types.js";
import { checkCpuAttestation as checkCpuAttestationDefault } from "./cpu.js";
import { checkNvidiaGpuAttestation as checkNvidiaGpuAttestationDefault } from "./nvidia.js";
import { checkProofOfCloud as checkProofOfCloud_ } from "./proofOfCloud.js";
import { verifyWorkload as verifyWorkloadDefault, extractDstackAppId, type DockerFilesInput } from "./workload.js";
import {
  classifyTlsBinding,
  endpointBaseUrl,
  extractDockerCompose,
  getTlsCertBinding as getTlsCertBindingDefault,
  hasExplicitPort,
  normalizeEndpointUrl,
  parseEndpointUrl,
  parseServiceBaseUrl,
} from "./url.js";

const SECRET_VM_PORT = 29343;
// When a bare host (no explicit port) is given, we probe SECRET_VM_PORT first
// (attest-rest bound directly, standard VMs) and fall back to this port, where
// a host-net Caddy fronts the app-cert TLS and proxies the evidence endpoints
// (jedi/rytn topology; attest-rest is loopback-only and unreachable on 29343).
const SECRET_VM_TLS_FALLBACK_PORT = 21434;
const PROBE_TIMEOUT_MS = 5000;
const HTTPS_PORT = 443;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

// Re-export for backwards compatibility
export { extractDockerCompose };

export type SecretVmEndpoint = {
  host: string;
  port: number;
  servername: string;
  pathPrefix: string;
  baseUrl: string;
};

export type CheckSecretVmOptions = {
  product?: string;
  reloadAmdKds?: boolean;
  checkProofOfCloud?: boolean;
  dockerFilesInput?: DockerFilesInput;
  strict?: boolean;
  tlsUrl?: string;
  enforceGpu?: boolean;
};

export type SecretVmRuntime = {
  fetch: typeof fetch;
  getTlsCertBinding: typeof getTlsCertBindingDefault;
  checkCpuAttestation: typeof checkCpuAttestationDefault;
  checkNvidiaGpuAttestation: typeof checkNvidiaGpuAttestationDefault;
  verifyWorkload: typeof verifyWorkloadDefault;
};

function defaultSecretVmRuntime(): SecretVmRuntime {
  return {
    fetch: globalThis.fetch,
    getTlsCertBinding: getTlsCertBindingDefault,
    checkCpuAttestation: checkCpuAttestationDefault,
    checkNvidiaGpuAttestation: checkNvidiaGpuAttestationDefault,
    verifyWorkload: verifyWorkloadDefault,
  };
}

function isExplicitUrl(value: string): boolean {
  const trimmed = value.trim();
  return /^[A-Za-z][A-Za-z0-9+.-]*:\/\//.test(trimmed);
}

export function isEndpointLike(value: string): boolean {
  const trimmed = value.trim();
  if (trimmed === "") return false;
  if (isExplicitUrl(trimmed)) return true;
  if (trimmed.startsWith("[") || trimmed.includes(":") || trimmed.includes("/")) {
    return true;
  }
  if (trimmed === "localhost") return true;
  if (/^\d{1,3}(?:\.\d{1,3}){3}$/.test(trimmed)) return true;
  return /^[A-Za-z0-9-]+(?:\.[A-Za-z0-9-]+)+$/.test(trimmed);
}

export function parseVmUrl(url: string): { host: string; port: number } {
  const parsed = parseEndpointUrl(url, SECRET_VM_PORT);
  return {
    host: parsed.host,
    port: parsed.port,
  };
}

function makeAttestationEndpoint(
  url: string,
  defaultPort: number = SECRET_VM_PORT,
): SecretVmEndpoint {
  const parsed = parseServiceBaseUrl(url, defaultPort, "Attestation URL");
  return {
    host: parsed.host,
    port: parsed.port,
    servername: parsed.host,
    pathPrefix: parsed.pathPrefix,
    baseUrl: endpointBaseUrl(parsed),
  };
}

function makeTlsEndpoint(url: string): SecretVmEndpoint {
  url = url.trim();
  if (url === "") {
    throw new Error("TLS binding URL must not be empty");
  }
  const normalizedUrl = normalizeEndpointUrl(url);
  const parsed = new URL(normalizedUrl);
  if (parsed.protocol !== "https:") {
    throw new Error("TLS binding URL must use https://");
  }
  const { host, port, username, password, pathPrefix, search, hash } = parseEndpointUrl(url, HTTPS_PORT);
  if (username !== "" || password !== "") {
    throw new Error("TLS binding URL must not include userinfo");
  }
  if (pathPrefix !== "") {
    throw new Error("TLS binding URL must not include a path");
  }
  if (search !== "") {
    throw new Error("TLS binding URL must not include a query string");
  }
  if (hash !== "") {
    throw new Error("TLS binding URL must not include a fragment");
  }
  return {
    host,
    port,
    servername: host,
    pathPrefix: "",
    baseUrl: endpointBaseUrl({
      host,
      port,
      protocol: "https:",
      username: "",
      password: "",
      pathPrefix: "",
      search: "",
      hash: "",
    }),
  };
}

export function resolveSecretVmEndpoints(
  attestationUrl: string,
  tlsUrl?: string,
): { attestation: SecretVmEndpoint; tls: SecretVmEndpoint } {
  const attestation = makeAttestationEndpoint(attestationUrl);
  return {
    attestation,
    tls: tlsUrl === undefined ? attestation : makeTlsEndpoint(tlsUrl),
  };
}

// probeCpu returns true if the endpoint answers GET /cpu within the timeout.
// A reachable /cpu marks the origin that serves both the quote and the
// app-cert TLS, so attestation and tls endpoints share that port.
async function probeCpu(
  endpoint: SecretVmEndpoint,
  fetchImpl: typeof fetch,
): Promise<boolean> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), PROBE_TIMEOUT_MS);
  try {
    const resp = await fetchImpl(`${endpoint.baseUrl}/cpu`, { signal: controller.signal });
    return resp.ok;
  } catch {
    return false;
  } finally {
    clearTimeout(timer);
  }
}

// resolveSecretVmEndpointsWithFallback mirrors resolveSecretVmEndpoints but,
// when the URL carries no explicit port and no separate tls URL was given,
// probes SECRET_VM_PORT (29343) then SECRET_VM_TLS_FALLBACK_PORT (21434) and
// binds both endpoints to whichever answers /cpu. An explicit port or an
// explicit tls URL disables probing and preserves the exact prior behavior.
async function resolveSecretVmEndpointsWithFallback(
  attestationUrl: string,
  tlsUrl: string | undefined,
  fetchImpl: typeof fetch,
): Promise<{ attestation: SecretVmEndpoint; tls: SecretVmEndpoint }> {
  if (tlsUrl !== undefined || hasExplicitPort(attestationUrl)) {
    return resolveSecretVmEndpoints(attestationUrl, tlsUrl);
  }
  const candidatePorts = [SECRET_VM_PORT, SECRET_VM_TLS_FALLBACK_PORT];
  let first: SecretVmEndpoint | undefined;
  for (const port of candidatePorts) {
    const endpoint = makeAttestationEndpoint(attestationUrl, port);
    if (first === undefined) first = endpoint;
    if (await probeCpu(endpoint, fetchImpl)) {
      return { attestation: endpoint, tls: endpoint };
    }
  }
  // None reachable: fall back to the primary port so downstream fetch surfaces
  // a clear error against 29343.
  return { attestation: first!, tls: first! };
}

// ---------------------------------------------------------------------------
// Public
// ---------------------------------------------------------------------------

export async function checkSecretVm(
  url: string,
  options?: CheckSecretVmOptions,
): Promise<AttestationResult>;
export async function checkSecretVm(
  url: string,
  product?: string,
  reloadAmdKds?: boolean,
  checkProofOfCloud?: boolean,
  dockerFilesInput?: DockerFilesInput,
  strict?: boolean,
  tlsUrl?: string,
  enforceGpu?: boolean,
): Promise<AttestationResult>;
export async function checkSecretVm(
  url: string,
  productOrOptions: string | CheckSecretVmOptions = "",
  reloadAmdKds = false,
  checkProofOfCloud = false,
  dockerFilesInput?: DockerFilesInput,
  strict = false,
  tlsUrl?: string,
  enforceGpu = false,
): Promise<AttestationResult> {
  let options: CheckSecretVmOptions;

  if (typeof productOrOptions === "object") {
    options = productOrOptions;
  } else {
    if (productOrOptions !== "" && isEndpointLike(productOrOptions)) {
      return makeResult("SECRET-VM", {
        errors: [
          `Endpoint-like second argument ${JSON.stringify(productOrOptions)} is not a product name. ` +
            "Pass split service endpoints with checkSecretVm(url, { tlsUrl: ... }).",
        ],
      });
    }
    options = {
      product: productOrOptions,
      reloadAmdKds,
      checkProofOfCloud,
      dockerFilesInput,
      strict,
      tlsUrl,
      enforceGpu,
    };
  }

  return checkSecretVmWithRuntime(url, options, defaultSecretVmRuntime());
}

export async function checkSecretVmWithRuntime(
  url: string,
  options: CheckSecretVmOptions,
  runtime: SecretVmRuntime,
): Promise<AttestationResult> {
  const errors: string[] = [];
  const checks: Record<string, boolean> = {};
  const report: Record<string, any> = {};
  const product = options.product ?? "";
  const reloadAmdKds = options.reloadAmdKds ?? false;
  const checkProofOfCloud = options.checkProofOfCloud ?? false;
  const dockerFilesInput = options.dockerFilesInput;
  const strict = options.strict ?? false;
  const enforceGpu = options.enforceGpu ?? false;

  let endpoints: ReturnType<typeof resolveSecretVmEndpoints>;
  try {
    endpoints = await resolveSecretVmEndpointsWithFallback(url, options.tlsUrl, runtime.fetch);
  } catch (e: any) {
    return makeResult("SECRET-VM", { errors: [e.message] });
  }

  const baseUrl = endpoints.attestation.baseUrl;
  report.attestation_url = endpoints.attestation.baseUrl;
  report.tls_binding_url = endpoints.tls.baseUrl;

  // 1. TLS certificate digests (SPKI + full certificate, for backward compat)
  let tlsBinding: Awaited<ReturnType<typeof getTlsCertBindingDefault>>;
  try {
    tlsBinding = await runtime.getTlsCertBinding(
      endpoints.tls.host,
      endpoints.tls.port,
      endpoints.tls.servername,
    );
    checks.tls_cert_fetched = true;
    report.tls_spki_fingerprint = tlsBinding.spki.toString("hex");
    report.tls_certificate_fingerprint = tlsBinding.certificate.toString("hex");
  } catch (e: any) {
    errors.push(`Failed to get TLS certificate: ${e.message}`);
    checks.tls_cert_fetched = false;
    return makeResult("SECRET-VM", { checks: orderChecks(checks), report, errors });
  }

  // 2. Fetch and verify CPU quote
  let cpuData: string;
  try {
    const resp = await runtime.fetch(`${baseUrl}/cpu`);
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    cpuData = await resp.text();
    checks.cpu_quote_fetched = true;
  } catch (e: any) {
    errors.push(`Failed to fetch CPU quote: ${e.message}`);
    checks.cpu_quote_fetched = false;
    return makeResult("SECRET-VM", { checks: orderChecks(checks), report, errors });
  }

  const cpuResult = await runtime.checkCpuAttestation(cpuData, product, reloadAmdKds, strict);
  checks.cpu_quote_verified = cpuResult.valid;
  report.cpu = cpuResult.report;
  report.cpu_type = cpuResult.attestationType;
  if (!cpuResult.valid) {
    errors.push(...cpuResult.errors);
  }

  // 3. TLS binding: first 32 bytes of report_data == SHA-256(SPKI DER) [current]
  //    or SHA-256(full certificate DER) [legacy]. Accept either so a mixed fleet
  //    (SPKI-pinned + older full-cert VMs) keeps verifying.
  const reportDataHex: string = cpuResult.report.report_data ?? "";
  if (reportDataHex.length >= 64) {
    const firstHalf = reportDataHex.slice(0, 64);
    const binding = classifyTlsBinding(firstHalf, tlsBinding);
    checks.tls_binding_verified = binding.verified;
    if (binding.verified) {
      report.tls_binding_kind = binding.kind;
    } else {
      errors.push(
        `TLS binding failed: report_data first half (${firstHalf.slice(0, 16)}...) ` +
          `!= TLS SPKI (${tlsBinding.spki.toString("hex").slice(0, 16)}...) ` +
          `or certificate (${tlsBinding.certificate.toString("hex").slice(0, 16)}...) digest`,
      );
    }
  } else {
    checks.tls_binding_verified = false;
    errors.push("report_data too short for TLS binding check");
  }

  // 4. Fetch GPU quote (optional). GPU is only required when enforceGpu is set.
  let gpuPresent = false;
  let gpuData = "";
  try {
    const resp = await runtime.fetch(`${baseUrl}/gpu`);
    if (resp.ok) {
      gpuData = await resp.text();
      const parsed = JSON.parse(gpuData);
      if ("error" in parsed) {
        checks.gpu_quote_fetched = false;
      } else {
        gpuPresent = true;
        checks.gpu_quote_fetched = true;
      }
    } else {
      checks.gpu_quote_fetched = false;
    }
  } catch {
    checks.gpu_quote_fetched = false;
  }

  if (gpuPresent) {
    const gpuResult = await runtime.checkNvidiaGpuAttestation(gpuData);
    checks.gpu_quote_verified = gpuResult.valid;
    report.gpu = gpuResult.report;
    if (!gpuResult.valid) {
      errors.push(...gpuResult.errors);
    }

    // 5. GPU binding: second 32 bytes of report_data == GPU nonce
    const gpuJson = JSON.parse(gpuData);
    const gpuNonce: string = gpuJson.nonce ?? "";
    if (reportDataHex.length >= 128) {
      const secondHalf = reportDataHex.slice(64, 128);
      checks.gpu_binding_verified = secondHalf === gpuNonce;
      if (!checks.gpu_binding_verified) {
        errors.push(
          `GPU binding failed: report_data second half (${secondHalf.slice(0, 16)}...) ` +
            `!= GPU nonce (${gpuNonce.slice(0, 16)}...)`,
        );
      }
    } else {
      checks.gpu_binding_verified = false;
      errors.push("report_data too short for GPU binding check");
    }
  }

  // 5b. GPU enforcement (opt-in): when enforceGpu is set, a GPU must be present,
  // so a CPU-only VM fails closed instead of silently passing.
  if (enforceGpu) {
    checks.gpu_present = gpuPresent;
    if (!gpuPresent) {
      errors.push(
        "GPU attestation required (--enforce-gpu) but this VM exposes no GPU",
      );
    }
  }

  // 6. Fetch and verify workload (docker-compose)
  try {
    const resp = await runtime.fetch(`${baseUrl}/docker-compose`);
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    const dockerCompose = await resp.text();
    checks.workload_fetched = true;

    // Newer (dstack) VMs measure the app-id as RTMR3's first event. Fetch it
    // from /info (best-effort — older images have no /info, so we fall back to
    // the pre-dstack schema and never fail the check on a missing endpoint).
    let dstackAppId = "";
    try {
      const infoResp = await runtime.fetch(`${baseUrl}/info`);
      if (infoResp.ok) dstackAppId = extractDstackAppId(await infoResp.text());
    } catch {
      /* no /info endpoint — old schema */
    }
    const workloadResult = await runtime.verifyWorkload(cpuData, dockerCompose, dockerFilesInput, dstackAppId);
    checks.workload_binding_verified = workloadResult.status === "authentic_match";
    // The app-id is only *proven* when it was an input to a TDX RTMR3 replay
    // that reproduced a hardware-signed quote. SEV-SNP has no app-id in its
    // launch measurement, and a failed TDX replay proves nothing either — in
    // both cases the value is whatever the VM chose to serve on /info. The
    // cpuResult.valid conjunct matters because verification does not stop at a
    // failed CPU quote: verifyTdxWorkload replays measurements without checking
    // the DCAP signature, so an unsigned quote carrying copied measurements can
    // still reach authentic_match. Report the value either way (it is useful
    // for diagnosis) but never without saying which case it is.
    if (dstackAppId) {
      report.dstack_app_id = dstackAppId;
      report.dstack_app_id_verified =
        cpuResult.valid &&
        report.cpu_type === "TDX" &&
        workloadResult.status === "authentic_match";
    }
    report.workload = workloadResult;
    report.docker_compose = dockerCompose;
    if (workloadResult.status === "authentic_mismatch") {
      errors.push("Workload mismatch: VM is authentic but docker-compose does not match");
    } else if (workloadResult.status === "not_authentic") {
      errors.push("Workload verification failed: not an authentic SecretVM");
    }
  } catch (e: any) {
    errors.push(`Failed to fetch workload: ${e.message}`);
    checks.workload_fetched = false;
  }

  // 7. Proof of cloud (opt-in): the community trust-server peers confirm the
  // machine is on the Proof of Cloud whitelist. Disabled by default — pass
  // checkProofOfCloud=true (or --proof-of-cloud on the CLI) to include this check.
  if (checkProofOfCloud) {
    const pocResult = await checkProofOfCloud_(cpuData);
    checks.proof_of_cloud_verified = pocResult.valid;
    if (pocResult.report.proof_of_cloud !== undefined) {
      report.proof_of_cloud = pocResult.report.proof_of_cloud;
    }
    if (!pocResult.valid) {
      errors.push(...pocResult.errors);
    }
  }

  // Overall validity
  const requiredChecks = [
    checks.tls_cert_fetched,
    checks.cpu_quote_fetched,
    checks.cpu_quote_verified,
    checks.tls_binding_verified,
    !!checks.workload_binding_verified,
  ];
  if (checkProofOfCloud) {
    requiredChecks.push(!!checks.proof_of_cloud_verified);
  }
  if (enforceGpu) {
    requiredChecks.push(!!checks.gpu_present);
  }
  if (gpuPresent) {
    requiredChecks.push(!!checks.gpu_quote_verified);
    requiredChecks.push(!!checks.gpu_binding_verified);
  }
  const valid = requiredChecks.every(Boolean);

  return makeResult("SECRET-VM", { valid, checks: orderChecks(checks), report, errors });
}

File: python/src/secretvm/verify/vm.py
"""Secret VM end-to-end verification (CPU + GPU + binding checks)."""

import asyncio
import hashlib
import json
import re
import socket
import ssl
import sys
from dataclasses import dataclass
from typing import Optional
from urllib.parse import unquote, urlparse

import requests
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
from cryptography.x509 import load_der_x509_certificate

from .types import AttestationResult, order_checks

_SECRET_VM_PORT = 29343
_SECRET_VM_RESOURCE_PATHS = {"cpu", "gpu", "docker-compose"}


@dataclass(frozen=True)
class _Endpoint:
    host: str
    port: int
    path_prefix: str
    base_url: str


def _get_tls_cert_digests(host: str, port: int) -> tuple[bytes, bytes]:
    """Connect to host:port and return (SHA-256 of SPKI DER, SHA-256 of full cert DER).

    The SPKI digest is the current binding (stable across cert renewals); the full
    certificate digest is the legacy binding. Accepting either keeps a mixed fleet
    verifiable. The platform trust store is enforced (create_default_context) before
    hashing anything.
    """
    context = ssl.create_default_context()
    with socket.create_connection((host, port), timeout=10) as sock:
        with context.wrap_socket(sock, server_hostname=host) as tls_sock:
            der = tls_sock.getpeercert(binary_form=True)
    if not der:
        raise ssl.SSLError("No certificate received")
    cert = load_der_x509_certificate(der)
    spki = cert.public_key().public_bytes(Encoding.DER, PublicFormat.SubjectPublicKeyInfo)
    return hashlib.sha256(spki).digest(), hashlib.sha256(der).digest()


def _get_tls_cert_fingerprint(host: str, port: int) -> bytes:
    """Return SHA-256 of the certificate SPKI DER (kept for backward compatibility)."""
    return _get_tls_cert_digests(host, port)[0]


def _classify_tls_binding(first_half_hex: str, spki: bytes, cert: bytes):
    """Return (verified, kind) for report_data's first 32 bytes against either digest.

    kind is "spki" (current) or "certificate" (legacy); None when unverified.
    """
    if first_half_hex == spki.hex():
        return True, "spki"
    if first_half_hex == cert.hex():
        return True, "certificate"
    return False, None


def _format_url_host(host: str) -> str:
    return f"[{host}]" if ":" in host and not host.startswith("[") else host


def _normalize_endpoint(endpoint: str) -> str:
    endpoint = endpoint.strip()
    if "://" not in endpoint:
        endpoint = f"https://{endpoint}"
    return endpoint


def _path_prefix(path: str) -> str:
    trimmed = path.rstrip("/")
    return "" if trimmed in ("", "/") else trimmed


def _decode_path_segments(path_prefix: str, label: str) -> list[str]:
    segments = [segment for segment in path_prefix.split("/") if segment]
    decoded = []
    for segment in segments:
        if re.search(r"%(?![0-9A-Fa-f]{2})", segment):
            raise ValueError(f"{label} contains invalid percent-encoding")
        item = unquote(segment, errors="strict")
        if "/" in item or "\\" in item:
            raise ValueError(f"{label} must not contain encoded path separators")
        decoded.append(item)
    return decoded


def _parse_service_base_url(
    endpoint: str,
    default_port: int = _SECRET_VM_PORT,
    label: str = "SecretVM endpoint URL",
) -> _Endpoint:
    normalized = _normalize_endpoint(endpoint)
    parsed = urlparse(normalized)
    if parsed.scheme != "https":
        raise ValueError(f"{label} must use https://")
    if parsed.username or parsed.password:
        raise ValueError(f"{label} must not include userinfo")
    if parsed.query:
        raise ValueError(f"{label} must not include a query string")
    if parsed.fragment:
        raise ValueError(f"{label} must not include a fragment")
    if not parsed.hostname:
        raise ValueError(f"{label} must include a host")

    path_prefix = _path_prefix(parsed.path)
    decoded_segments = _decode_path_segments(path_prefix, label)
    if decoded_segments and decoded_segments[-1] in _SECRET_VM_RESOURCE_PATHS:
        raise ValueError(
            f"{label} must be a service base URL, not a concrete /{decoded_segments[-1]} resource path"
        )

    try:
        parsed_port = parsed.port
    except ValueError as e:
        raise ValueError(f"{label} has an invalid port") from e
    port = parsed_port if parsed_port is not None else default_port
    if port < 1 or port > 65535:
        raise ValueError(f"{label} has an invalid port")
    base_url = f"https://{_format_url_host(parsed.hostname)}:{port}{path_prefix}"
    return _Endpoint(host=parsed.hostname, port=port, path_prefix=path_prefix, base_url=base_url)


def _parse_tls_endpoint(endpoint: str, label: str = "TLS endpoint URL") -> _Endpoint:
    parsed = _parse_service_base_url(endpoint, default_port=443, label=label)
    if parsed.path_prefix:
        raise ValueError(f"{label} must not include a path")
    return parsed


def _parse_vm_url(url: str) -> tuple[str, int]:
    """Extract host and port from a URL, defaulting to port 29343."""
    parsed = _parse_service_base_url(url)
    return parsed.host, parsed.port


def _extract_docker_compose(raw: str) -> str:
    """Legacy helper for old HTML-wrapped compose endpoints."""
    import html
    import re

    text = raw.strip()
    m = re.search(r"<pre>(.*?)</pre>", text, re.DOTALL | re.IGNORECASE)
    if m:
        text = m.group(1)
    text = html.unescape(text)
    # Strip zero-width spaces and other invisible Unicode characters
    text = re.sub(r"[\u200b\u200c\u200d\ufeff]", "", text)
    return text


def _get_pkg():
    """Return the secretvm.verify package module for late-bound lookups.

    This allows tests to mock functions at the package level
    (e.g. ``patch('secretvm.verify._get_tls_cert_fingerprint', ...)``).
    """
    return sys.modules["secretvm.verify"]


def check_secret_vm(
    url: str,
    product: str = "",
    reload_amd_kds: bool = False,
    check_proof_of_cloud: bool = False,
    docker_files: Optional[bytes] = None,
    docker_files_sha256: Optional[str] = None,
    strict: bool = False,
    enforce_gpu: bool = False,
) -> AttestationResult:
    """Verify a Secret VM by fetching CPU and GPU attestation from its endpoints.

    Connects to the VM's attestation service base URL, fetches the CPU
    quote from /cpu and the GPU quote from /gpu, verifies both,
    and checks the binding between:
      - The TLS certificate SPKI digest and the first half of report_data
      - The GPU nonce and the second half of report_data

    Args:
        url: VM attestation service base URL (e.g. "https://host:21434",
            "https://host:29343", or just "host").
        product: AMD product name (only used for SEV-SNP). Auto-detected if empty.
        reload_amd_kds: If True, bypass the local AMD KDS cache and re-fetch
            VCEK / cert chain / CRL. No effect on TDX VMs (TDX doesn't cache).

    Returns:
        AttestationResult with attestation_type="SECRET-VM".
    """
    # Use late-bound lookups so tests can mock at the package level
    pkg = _get_pkg()

    errors = []
    checks = {}
    report = {}

    try:
        endpoint = _parse_service_base_url(url)
    except Exception as e:
        errors.append(f"Invalid SecretVM endpoint URL: {e}")
        return AttestationResult(
            valid=False, attestation_type="SECRET-VM",
            checks=order_checks(checks), report=report, errors=errors,
        )

    host, port = endpoint.host, endpoint.port
    base_url = endpoint.base_url

    # 1. Get TLS certificate digests (SPKI + full certificate, for backward compat)
    try:
        tls_spki_fingerprint, tls_cert_fingerprint = pkg._get_tls_cert_digests(host, port)
        checks["tls_cert_fetched"] = True
        report["tls_spki_fingerprint"] = tls_spki_fingerprint.hex()
        report["tls_certificate_fingerprint"] = tls_cert_fingerprint.hex()
    except Exception as e:
        errors.append(f"Failed to get TLS certificate: {e}")
        checks["tls_cert_fetched"] = False
        return AttestationResult(
            valid=False, attestation_type="SECRET-VM",
            checks=order_checks(checks), report=report, errors=errors,
        )

    # 2. Fetch and verify CPU quote
    try:
        resp = pkg.requests.get(f"{base_url}/cpu", timeout=15, verify=True)
        resp.raise_for_status()
        cpu_data = resp.text
        checks["cpu_quote_fetched"] = True
    except Exception as e:
        errors.append(f"Failed to fetch CPU quote: {e}")
        checks["cpu_quote_fetched"] = False
        return AttestationResult(
            valid=False, attestation_type="SECRET-VM",
            checks=order_checks(checks), report=report, errors=errors,
        )

    cpu_result = pkg.check_cpu_attestation(
        cpu_data, product=product, reload_amd_kds=reload_amd_kds, strict=strict,
    )
    checks["cpu_quote_verified"] = cpu_result.valid
    report["cpu"] = cpu_result.report
    report["cpu_type"] = cpu_result.attestation_type
    if not cpu_result.valid:
        errors.extend(cpu_result.errors)

    # 3. Check TLS binding: first 32 bytes of report_data == SHA-256(SPKI DER)
    #    [current] or SHA-256(full certificate DER) [legacy]. Accept either.
    report_data_hex = cpu_result.report.get("report_data", "")
    if len(report_data_hex) >= 64:
        first_half = report_data_hex[:64]  # first 32 bytes as hex
        verified, kind = _classify_tls_binding(first_half, tls_spki_fingerprint, tls_cert_fingerprint)
        checks["tls_binding_verified"] = verified
        if verified:
            report["tls_binding_kind"] = kind
        else:
            errors.append(
                f"TLS binding failed: report_data first half ({first_half[:16]}...) "
                f"!= TLS SPKI ({tls_spki_fingerprint.hex()[:16]}...) "
                f"or certificate ({tls_cert_fingerprint.hex()[:16]}...) digest"
            )
    else:
        checks["tls_binding_verified"] = False
        errors.append("report_data too short for TLS binding check")

    # 4. Fetch and verify GPU quote (optional). GPU is only required when enforce_gpu is set.
    gpu_present = False
    gpu_data = ""
    try:
        resp = pkg.requests.get(f"{base_url}/gpu", timeout=15, verify=True)
        resp.raise_for_status()
        gpu_data = resp.json() if resp.headers.get("content-type", "").startswith("application/json") else json.loads(resp.text)
        if "error" in gpu_data:
            # Non-GPU machine: {"error": "...", "details": "..."}
            checks["gpu_quote_fetched"] = False
        else:
            gpu_present = True
            checks["gpu_quote_fetched"] = True
            gpu_data = resp.text  # pass raw JSON text to check_nvidia_gpu_attestation
    except Exception:
        checks["gpu_quote_fetched"] = False

    if gpu_present:
        gpu_result = pkg.check_nvidia_gpu_attestation(gpu_data)
        checks["gpu_quote_verified"] = gpu_result.valid
        report["gpu"] = gpu_result.report
        if not gpu_result.valid:
            errors.extend(gpu_result.errors)

        # 5. Check GPU binding: second 32 bytes of report_data == GPU nonce
        gpu_json = json.loads(gpu_data)
        gpu_nonce = gpu_json.get("nonce", "")
        if len(report_data_hex) >= 128:
            second_half = report_data_hex[64:128]  # second 32 bytes as hex
            checks["gpu_binding_verified"] = second_half == gpu_nonce
            if not checks["gpu_binding_verified"]:
                errors.append(
                    f"GPU binding failed: report_data second half ({second_half[:16]}...) "
                    f"!= GPU nonce ({gpu_nonce[:16]}...)"
                )
        else:
            checks["gpu_binding_verified"] = False
            errors.append("report_data too short for GPU binding check")

    # 5b. GPU enforcement (opt-in): when enforce_gpu is set, a GPU must be present,
    # so a CPU-only VM fails closed instead of silently passing.
    if enforce_gpu:
        checks["gpu_present"] = gpu_present
        if not gpu_present:
            errors.append(
                "GPU attestation required (--enforce-gpu) but this VM exposes no GPU"
            )

    # 6. Fetch and verify workload (docker-compose)
    try:
        resp = pkg.requests.get(f"{base_url}/docker-compose", timeout=15, verify=True)
        resp.raise_for_status()
        docker_compose = resp.text
        checks["workload_fetched"] = True

        workload_result = pkg.verify_workload(
            cpu_data, docker_compose, docker_files, docker_files_sha256,
        )
        checks["workload_binding_verified"] = workload_result.status == "authentic_match"
        report["workload"] = {
            "status": workload_result.status,
            "template_name": workload_result.template_name,
            "artifacts_ver": workload_result.artifacts_ver,
            "env": workload_result.env,
        }
        report["docker_compose"] = docker_compose
        if workload_result.status == "authentic_mismatch":
            errors.append("Workload mismatch: VM is authentic but docker-compose does not match")
        elif workload_result.status == "not_authentic":
            errors.append("Workload verification failed: not an authentic SecretVM")
    except Exception as e:
        errors.append(f"Failed to fetch workload: {e}")
        checks["workload_fetched"] = False

    # 7. Proof of cloud (opt-in): ask the community trust-server peers whether
    # this machine is on the Proof of Cloud whitelist. Disabled by default — set
    # check_proof_of_cloud=True (or --proof-of-cloud on the CLI) to include.
    if check_proof_of_cloud:
        poc_result = pkg.check_proof_of_cloud(cpu_data)
        checks["proof_of_cloud_verified"] = poc_result.valid
        if poc_result.report.get("proof_of_cloud") is not None:
            report["proof_of_cloud"] = poc_result.report["proof_of_cloud"]
        if not poc_result.valid:
            errors.extend(poc_result.errors)

    # Overall validity
    required_checks = [
        checks.get("tls_cert_fetched"),
        checks.get("cpu_quote_fetched"),
        checks.get("cpu_quote_verified"),
        checks.get("tls_binding_verified"),
        checks.get("workload_binding_verified", False),
    ]
    if check_proof_of_cloud:
        required_checks.append(checks.get("proof_of_cloud_verified", False))
    if enforce_gpu:
        required_checks.append(checks.get("gpu_present", False))
    if gpu_present:
        required_checks.append(checks.get("gpu_quote_verified"))
        required_checks.append(checks.get("gpu_binding_verified"))

    valid = all(required_checks)

    return AttestationResult(
        valid=valid, attestation_type="SECRET-VM",
        checks=order_checks(checks), report=report, errors=errors,
    )


async def check_secret_vm_async(
    url: str,
    product: str = "",
    reload_amd_kds: bool = False,
    check_proof_of_cloud: bool = False,
    docker_files: Optional[bytes] = None,
    docker_files_sha256: Optional[str] = None,
    strict: bool = False,
    enforce_gpu: bool = False,
) -> AttestationResult:
    """Async variant of :func:`check_secret_vm`.

    Use this from inside an event loop. The current implementation offloads
    the synchronous wrapper to a thread pool via :func:`asyncio.to_thread`,
    so the event loop is not blocked while the TLS, CPU, GPU, and workload
    fetches run sequentially in the worker thread. A future refactor could
    parallelize the four fetches for additional speedup, but the simple
    wrapper is sufficient to use the API from async contexts today.
    """
    return await asyncio.to_thread(
        check_secret_vm, url, product, reload_amd_kds, check_proof_of_cloud,
        docker_files, docker_files_sha256, strict, enforce_gpu,
    )
